pkg/controller/master/backup/backup.go

```diff
@@ -226,6 +247,65 @@ func (h *Handler) getBackupSource(vmBackup *harvesterv1.VirtualMachineBackup) (*
 	return sourceVM, nil
 }
 
+func (h *Handler) getBackupSourceInstance(vmBackup *harvesterv1.VirtualMachineBackup) (*kubevirtv1.VirtualMachineInstance, error) {
+	if vmBackup.Spec.Source.Kind != kubevirtv1.VirtualMachineGroupVersionKind.Kind {
+		return nil, fmt.Errorf("unsupported source: %+v", vmBackup.Spec.Source)
+	}
+
+	sourceVMI, err := h.vmisCache.Get(vmBackup.Namespace, vmBackup.Spec.Source.Name)
+	if err != nil {
+		return nil, err
+	}
+	if sourceVMI.DeletionTimestamp != nil {
+		return nil, fmt.Errorf("vm %s/%s is being deleted", vmBackup.Namespace, vmBackup.Spec.Source.Name)
+	}
+
+	return sourceVMI, nil
+}
+
+func (h *Handler) tryFreezeFS(backup *harvesterv1.VirtualMachineBackup) {
+	sourceVMI, err := h.getBackupSourceInstance(backup)
+	if err != nil {
+		logrus.WithError(err).Warn("vmi not found during backup")
+		return
+	}
+
+	hasAgent := false
+
+	for _, condition := range sourceVMI.Status.Conditions {
+		if condition.Type == kubevirtv1.VirtualMachineInstanceAgentConnected {
+			hasAgent = condition.Status == corev1.ConditionTrue
+			break
+		}
+	}
+
+	if !hasAgent {
+		return
+	}
+
+	if err := h.freezeFS(context.Background(), sourceVMI, defaultFreezeDuration); err != nil {
```

Comment:
May I have two questions,
1.  Can we guarantee one second is long enough for all CSI drivers (Longhorn currently) to take a snapshot? Since it may be related to complex operations, as Longhorn for example, the driver has to [create a new layer for replicas and update read indexes on all nodes ](https://longhorn.io/docs/1.4.2/concepts/#241-how-snapshots-work)
2. As stated above, if taking a snapshot requires longer than we expected, filesystem I/O could be back to work before snapshot completion, and it may still brings data inconsistency between the origin VM and snapshot. Should we consider freezing vm filesystem before taking a snapshot, and unfreezing it after the snapshot, like [KubeVirt's handling for snapshot](https://kubevirt.io/user-guide/operations/snapshot_restore_api/#snapshot-a-virtualmachine)? 

Thanks

Reply:
FYI:
I run Harvester HOST on KVM VM, and take vm snapshort/restore frequently in Virt-manager, to switch between diffrent Harvester versions, for such scale of VM (10 core, 20G memory, 200G disk), even in really fast PC, it still takes a long time to finish a snapshort, inbetween that, the VM is shown as suspend/freezed.

Back to here, seems difficult to set an one-fit-all timeout parameter, we may need to calculate it per VM scale and experience.

Or even expose the timeout as a parameter to `take snapshort`, give an initial estimated value, and user can further change it.

https://bugzilla.redhat.com/show_bug.cgi?id=1989263
https://github.com/kubevirt/kubevirt/pull/6219

Reply:
![image](https://github.com/harvester/harvester/assets/31133476/1eb923d4-8b2c-43d9-810f-35420f9250b2)

```
$ date
Wed 31 May 2023 03:40:41 PM CEST

$ date
Wed 31 May 2023 03:41:37 PM CEST
```

Did a test, this snapshot took about one minute. The VM is shown as `paused` in virt-manager.

Reply:
No, it does not block here.

This PR tries to freeze the fs for a moment (and thaw automatically), is an indepedent enhancement.

For the potential data crash/in-consistency, we need to further analyze other places, sure, which is out of scope of this PR. thanks.

Reply:
IMHO,

> > Can we guarantee one second is long enough for all CSI drivers (Longhorn currently) to take a snapshot? Since it may be related to complex operations, as Longhorn for example, the driver has to [create a new layer for replicas and update read indexes on all nodes ](https://longhorn.io/docs/1.4.2/concepts/#241-how-snapshots-work)
> 
> No, I don't think so.
> 
> I think the ideal freeze duration is not the duration of the snapshot, but instead however long it takes to flush all I/O that was pending at the time the user executed the snapshot. I believe that duration is much, much smaller than the duration of a whole snapshot. Maybe 1 second is too low, but the duration of the snapshot would be the worst case, IMO. I'll comment more on this below.
> 
> > As stated above, if taking a snapshot requires longer than we expected, filesystem I/O could be back to work before snapshot completion, and it may bring unexpected behaviour.
> 
> I think this is also true for how snapshots/backups in Harvester work today.
> 
> I don't know anything about Longhorn internals, but ideally during a snapshot, it would be redirecting new writes from the guest to some sort of overlay that can be merged down to the original disk after the snapshot completes. thinking If this isn't how it behaves for snapshots then indeed this complicates things.

From my understanding, Longhorn has to create a new layer for all replicas and update the read indexes for all nodes. If during this process, the filesystem is thawed and back to process I/Os again. The snapshotted storage status could be uncertain.  And different CSI drivers may have different behaviors?

> 
> > Should we consider freezing vm filesystem before taking a snapshot, and unfreezing it after the snapshot, like [KubeVirt's handling for snapshot](https://kubevirt.io/user-guide/operations/snapshot_restore_api/#snapshot-a-virtualmachine)?
> 
> I would prefer to freeze for exactly how long it takes to flush pending I/O to disk at time of snapshot vs freezing for the duration of the snapshot, which I suspect will take much longer than flushing the pending I/O.
> 
> I might be misunderstanding how the FS freezing mechanism works, but if the file system is frozen, then none of the guest's writes are making it to durable storage. So, for the window that the FS is frozen, the risk of dataloss is greater in case of an unexpected shutdown from the guest, which is why I'd prefer to freeze for as small of a duration as possible vs the duration of the entire snapshot.

According to:
- https://github.com/kubevirt/kubevirt/blob/main/docs/freeze.md "Kubevirt provides a way to freeze/thaw the guest filesystem. That can be used to ensure the filesystem consistency when making a backup or a VM snapshot." 
- https://linux.die.net/man/8/fsfreeze "all dirty data, metadata, and log information are written to disk"

I think filesystem freeze is here to guarantee data consistency, but like what you pointed, the new coming I/Os would be halted

> 
> That said, freezing the FS for exactly the amount of time it takes to flush pending I/O is a very hand-wavey thing to say, and probably the best thing we can do is make it tunable.

Agree, making timeout tunable seems a practical solution. [The default deadline for KubeVirt vmSnapshot is 5 minutes](https://kubevirt.io/user-guide/operations/snapshot_restore_api/#snapshot-a-virtualmachine), and if the vmSnapshot has not successfully completed before the deadline, it will be marked as Failed. 
> 
> > I run Harvester HOST on KVM VM, and take vm snapshort/restore frequently in Virt-manager, to switch between diffrent Harvester versions, for such scale of VM (10 core, 20G memory, 200G disk), even in really fast PC, it still takes a long time to finish a snapshort, inbetween that, the VM is shown as suspend/freezed.
> 
> Yeah, I think in this case the "suspend/freeze" isn't referring to the FS being frozen, but instead libvirt pausing the domain to take a full system snapshot (cpu registers, RAM, disk), resulting in downtime. The alternative to this is having libvirt do an external disk snapshot where it just snapshots the disk while it redirects new writes from the live guest to a scratch file and then merges that back into the original disk once the snapshot completes. (Sorry if none of this is new info, I just want to be careful where I use the word "freeze" in this PR so I can be clear which mechanism I'm referring to).
> 
> Using the qemu-guest-agent in Harvester as shown won't pause the VM, so the VM should not experience downtime in that way.
> 
> > Back to here, seems difficult to set an one-fit-all timeout parameter, we may need to calculate it per VM scale and experience.
> > Or even expose the timeout as a parameter to take snapshort, give an initial estimated value, and user can further change it.
> 
> I think exposing the parameter is probably the most user-friendly thing here. We can have a conservative default of a few seconds, but users can choose to extend this if they want, based on the size of their VM or how I/O intensive their workload is.
> 
> Is this PR blocked on that parameter?